### PR TITLE
bpo-35153: Add headers to xmlrpc.client.ServerProxy

### PR DESCRIPTION
**Pull request by cedk**
Allow to add headers to xmlrpc requests sent to the server.

https://bugs.python.org/issue35153

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: Doc/library/xmlrpc.client.rst [hunk 1]
```diff
@@ -61,6 +61,8 @@ between conformable Python objects and XML on the wire.
    may be passed to calls.
    The obsolete *use_datetime* flag is similar to *use_builtin_types* but it
    applies only to date/time values.
+   The *headers* list can be used to set HTTP headers to each request. It is
```
Comment (matrixise) [inline Doc/library/xmlrpc.client.rst]: Could you add a `..versionadded:: headers` with the description.

File: Misc/NEWS.d/next/Library/2018-11-03-12-38-03.bpo-35153.009pdF.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Add headers to xmlrpc.client.ServerProxy
```
Comment (matrixise) [inline Misc/NEWS.d/next/Library/2018-11-03-12-38-03.bpo-35153.009pdF.rst]: You could add "Patch by Cédric Krier"

**Review by matrixise — CHANGES_REQUESTED** [review]
You have to add some tests for the headers in `Lib/test/test_xmlrpc.py`.

Thank you

File: Doc/library/xmlrpc.client.rst [hunk 3]
```diff
@@ -34,7 +34,7 @@ between conformable Python objects and XML on the wire.
 
 .. class:: ServerProxy(uri, transport=None, encoding=None, verbose=False, \
                        allow_none=False, use_datetime=False, \
-                       use_builtin_types=False, *, context=None)
+                       use_builtin_types=False, headers=None, *, context=None)
```
Comment (serhiy-storchaka) [inline Doc/library/xmlrpc.client.rst]: In case of large number of arguments it is better to make new arguments keyword-only.

File: Lib/xmlrpc/client.py [hunk 4]
```diff
@@ -1131,10 +1131,12 @@ class Transport:
     # that they can decode such a request
     encode_threshold = None #None = don't encode
 
-    def __init__(self, use_datetime=False, use_builtin_types=False):
+    def __init__(self, use_datetime=False, use_builtin_types=False,
+                 headers=None):
```
Comment (serhiy-storchaka) [inline Lib/xmlrpc/client.py]: Keyword-only.

File: Lib/test/test_xmlrpc.py [hunk 5]
```diff
@@ -1171,6 +1171,26 @@ def test_gzip_decode_limit(self):
...
 
 
+class HeadersServerTestCase(BaseServerTestCase):
+    class RequestHandler(xmlrpc.server.SimpleXMLRPCRequestHandler):
+        test_header = None
+
+        def do_POST(self):
+            self.__class__.test_header = self.headers.get('x-test')
+            return super().do_POST()
+    requestHandler = RequestHandler
```
Comment (methane) [inline Lib/test/test_xmlrpc.py:1181]: Why this line is needed?

Comment (cedk) [reply Lib/test/test_xmlrpc.py:1181]: Because it is the attribute BaseServerTestCase is looking for in setUp to start the server.

**Review by methane — APPROVED** [review]
LGTM

File: Doc/library/xmlrpc.client.rst [hunk 6]
```diff
@@ -34,7 +34,7 @@ between conformable Python objects and XML on the wire.
...
 
    .. versionchanged:: 3.3
```
Comment (vstinner) [inline Doc/library/xmlrpc.client.rst]: Please move this "versionchanged 3.3" at the end, as we do in other parts of the doc.

File: Doc/library/xmlrpc.client.rst [hunk 7]
```diff
@@ -62,6 +62,10 @@ between conformable Python objects and XML on the wire.
    The obsolete *use_datetime* flag is similar to *use_builtin_types* but it
    applies only to date/time values.
```
Comment (vstinner) [inline Doc/library/xmlrpc.client.rst:63]: You must document header. Does it have to be a list? Can it be a set? Or any iterable?

File: Misc/NEWS.d/next/Library/2018-11-03-12-38-03.bpo-35153.009pdF.rst [hunk 8]
```diff
@@ -0,0 +1,2 @@
+Add ``headers`` to ``xmlrpc.client.ServerProxy``.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2018-11-03-12-38-03.bpo-35153.009pdF.rst]: You also modified Transport, no? Please mention it.

I don't understand why Transport isn't document at https://docs.python.org/dev/library/xmlrpc.client.html whereas it's used in an example. Maybe you want to write a separated change to document Transport class?

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Misc/NEWS.d/next/Library/2018-11-03-12-38-03.bpo-35153.009pdF.rst (hunk 8 above)
Comment (cedk) [reply Misc/NEWS.d/next/Library/2018-11-03-12-38-03.bpo-35153.009pdF.rst]: There is a comment in the doc about not everything is documented yet.
I may try to write it later.

File: Doc/library/xmlrpc.client.rst (hunk 7 above)
Comment (cedk) [reply Doc/library/xmlrpc.client.rst:63]: It is said in line 66 that it is a list.

Comment (vstinner) [reply Doc/library/xmlrpc.client.rst:63]: > It is said in line 66 that it is a list.

I would prefer to not document the parameter in the "versionchanged" markup, but in the "body" of the doc. versionchanged should basically say "new parameter".

You used. I would prefer to use versionchanged instead.

Just move the content of your versionadded into the body, and add something like "versionchanged:: 3.8 New *headers* parameter." (I let you format it properly ;-))

File: Doc/library/xmlrpc.client.rst [hunk 9]
```diff
@@ -34,10 +34,7 @@ between conformable Python objects and XML on the wire.
 
 .. class:: ServerProxy(uri, transport=None, encoding=None, verbose=False, \
                        allow_none=False, use_datetime=False, \
-                       use_builtin_types=False, *, context=None)
-
-   .. versionchanged:: 3.3
-      The *use_builtin_types* flag was added.
+                       use_builtin_types=False, headers=None, *, context=None)
```
Comment (pganssle) [inline Doc/library/xmlrpc.client.rst]: Does `headers` need to be a positional argument here? It seems more like `context` than like the other positional arguments to me.

File: Doc/library/xmlrpc.client.rst [hunk 10]
```diff
@@ -59,9 +56,17 @@ between conformable Python objects and XML on the wire.
    presented as :class:`bytes` objects; this flag is false by default.
    :class:`datetime.datetime`, :class:`bytes` and :class:`bytearray` objects
    may be passed to calls.
+   The *headers* list can be used to set HTTP headers to each request. It is
+   composed of tuple of header name and value.
```
Comment (pganssle) [inline Doc/library/xmlrpc.client.rst]: If I understand correctly, this is supposed to be a `list` of `(name, value)` pairs? If so, I would reword it something more like this:

```rst
The *headers* parameter is an optional list of HTTP headers to send
with each request, expressed as a :class:`list` of :class:`tuple`s representing
the header name and value. (e.g. `[('HeaderName', 'value')]`).
```

File: Lib/test/test_xmlrpc.py [hunk 11]
```diff
@@ -1171,6 +1171,26 @@ def test_gzip_decode_limit(self):
...
+
+    def setUp(self):
+        self.RequestHandler.test_header = None
+        return super().setUp()
+
+    def test_header(self):
```
Comment (pganssle) [inline Lib/test/test_xmlrpc.py:1197]: I can't say I really understand this test very well, but I do think I'd add other test cases, specifically the cases of setting more than one header and the case of an empty list.

Might also want to test what happens when you pass something other than a list, like a `tuple` or a `dict` for whatever behavior you expect. What about a `dict_items` object (e.g. `{'X-Test': 'foo'}.items()`)?

**Review by pganssle — CHANGES_REQUESTED** [review]

File: Doc/library/xmlrpc.client.rst (hunk 9 above)
Comment (serhiy-storchaka) [inline Doc/library/xmlrpc.client.rst]: And I would make the default value an empty tuple.

Comment (serhiy-storchaka) [inline Doc/library/xmlrpc.client.rst]:
```suggestion
                       use_builtin_types=False, *, context=None, headers=())
```

Comment (cedk) [reply Doc/library/xmlrpc.client.rst]: context is used only for https while headers is always used.

File: Lib/xmlrpc/client.py [hunk 12]
```diff
@@ -1131,10 +1131,12 @@ class Transport:
...
         self._use_datetime = use_datetime
         self._use_builtin_types = use_builtin_types
         self._connection = (None, None)
+        self._headers = headers or []
```
Comment (serhiy-storchaka) [inline Lib/xmlrpc/client.py]:
```suggestion
        self._headers = list(headers)
```

File: Lib/xmlrpc/client.py [hunk 13]
```diff
@@ -1265,7 +1267,7 @@ def close(self):
 
     def send_request(self, host, handler, request_body, debug):
         connection = self.make_connection(host)
-        headers = self._extra_headers[:]
+        headers = self._headers[:] + self._extra_headers[:]
```
Comment (serhiy-storchaka) [inline Lib/xmlrpc/client.py]:
```suggestion
        headers = self._headers + self._extra_headers
```

File: Lib/xmlrpc/client.py (hunk 4 above)
Comment (serhiy-storchaka) [inline Lib/xmlrpc/client.py]:
```suggestion
                 *, headers=()):
```

File: Misc/NEWS.d/next/Library/2018-11-03-12-38-03.bpo-35153.009pdF.rst [hunk 14]
```diff
@@ -0,0 +1,3 @@
+Add `headers` to :class:`xmlrpc.client.ServerProxy`,
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2018-11-03-12-38-03.bpo-35153.009pdF.rst]:
```suggestion
Add *headers* optional keyword-only parameter to :class:`xmlrpc.client.ServerProxy`,
```

**Review by vstinner — APPROVED** [review]
LGTM. It's now better documented and safer thanks to keyword-only parameters!

I just have proposed a minor change on the NEWS entry.

**Comment by vstinner** [comment]
Since @serhiy-storchaka and @pganssle asked for changes, I wait for them to review the updated PR.

File: Doc/library/xmlrpc.client.rst [hunk 15]
```diff
@@ -59,9 +56,18 @@ between conformable Python objects and XML on the wire.
    presented as :class:`bytes` objects; this flag is false by default.
    :class:`datetime.datetime`, :class:`bytes` and :class:`bytearray` objects
    may be passed to calls.
+   The *headers* parameter is an optional sequence of HTTP headers to send with
+   each request, expressed as a sequence of :class:`tuple` objects representing
```
Comment (serhiy-storchaka) [inline Doc/library/xmlrpc.client.rst]: I would use "of pairs" or "of 2-tuples".

File: Misc/NEWS.d/next/Library/2018-11-03-12-38-03.bpo-35153.009pdF.rst [hunk 16]
```diff
@@ -0,0 +1,3 @@
+Add *headers* optional keyword-only parameter to
+:class:`xmlrpc.client.ServerProxy`, :class:`xmlrpc.client.Transport` and
+:class:`xmlrpc.client.SafeTransport`.  Patch by Cédric Krier
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2018-11-03-12-38-03.bpo-35153.009pdF.rst]:
```suggestion
:class:`xmlrpc.client.SafeTransport`.  Patch by Cédric Krier.
```

File: Doc/library/xmlrpc.client.rst [hunk 17]
```diff
@@ -59,9 +56,18 @@ between conformable Python objects and XML on the wire.
...
+   the header name and value. (e.g. `[('Header-Name', 'value')]`).
    The obsolete *use_datetime* flag is similar to *use_builtin_types* but it
    applies only to date/time values.
 
+.. versionchanged:: 3.3
+    The *use_builtin_types* flag was added.
+
+.. versionchanged:: 3.8
+    The *headers* argument was added.
```
Comment (serhiy-storchaka) [inline Doc/library/xmlrpc.client.rst]:
```suggestion
    The *headers* parameter was added.
```

**Review by serhiy-storchaka — COMMENTED** [review]
"argument" is what passed when call a function. "parameter" is what specified in function declaration and used in function.

**Comment by serhiy-storchaka** [comment]
LGTM except few nitpicks.

File: Lib/test/test_xmlrpc.py [hunk 18]
```diff
@@ -1171,6 +1171,56 @@ def test_gzip_decode_limit(self):
...
 
 
+class HeadersServerTestCase(BaseServerTestCase):
+    class RequestHandler(xmlrpc.server.SimpleXMLRPCRequestHandler):
+        test_header = None
+        test_second_header = None
+
+        def do_POST(self):
+            self.__class__.test_header = self.headers.get('x-test')
+            self.__class__.test_second_header = self.headers.get(
```
Comment (pganssle) [inline Lib/test/test_xmlrpc.py]: Can you clarify why you need to stash these specifically-named headers on the class like this? Is it because these objects are ephemeral and you don't ever have direct access to them?

In any case, I think it would be preferable to just store the entire list of headers and check that it's what you expect, e.g.:

```python
def do_POST(self):
    self.__class__.headers_posted = copy.deepcopy(self.headers)
    super().do_POST()
```

Then the tests would look like this:

```python
p = xmlrpclib.ServerProxy(URL, headers=[('X-Test', 'foo')])

self.assertEqual(p.pow(6, 8), 6**8)
self.assertEqual(self.RequestHandler.headers_posted, {'X-Test': 'foo'})
```

I think this would be a stronger test, because it not only tests that the header was received, but also that no additional, anomalous headers were added, and it will prevent anyone from accidentally modifying the tests so that the test headers values aren't being cleared between test runs.

Comment (cedk) [reply Lib/test/test_xmlrpc.py]: Well headers are HTTPMessage so they can not be compared with dict. But I can test the keys of the headers.

File: Lib/test/test_xmlrpc.py [hunk 19]
```diff
@@ -1171,6 +1171,70 @@ def test_gzip_decode_limit(self):
...
+
+    def test_header_many(self):
+        p = xmlrpclib.ServerProxy(
+            URL, headers=[('X-Test', 'foo'), ('X-Test-Second', 'bar')])
+        self.assertEqual(p.pow(6, 8), 6**8)
+
+        headers = self.RequestHandler.test_headers
+        self.assertListEqual(
+            sorted(headers.keys()),
+            sorted(self.standard_headers + ['X-Test', 'X-Test-Second']))
```
Comment (pganssle) [inline Lib/test/test_xmlrpc.py]: I think maybe the best thing to do is to factor this out into a separate assertion function:

```python
def assertContainsAdditionalHeaders(self, headers, additional):
    expected_keys = sorted(self.standard_headers + list(additional.keys())
    self.assertListEqual(sorted(headers.keys()), expected_keys)

    for key in additional.keys():
        self.assertEqual(headers.get(key), additional[key])
```

Then you can use it everywhere like this:

```python
self.asssertContainsAdditionalHeaders(headers, {'X-Test': 'foo'})
```

Comment (cedk) [reply Lib/test/test_xmlrpc.py]: Indeed it is cleaner.

**Review by pganssle — APPROVED** [review]
OK, this looks good to me. Thanks for your contribution and your patience with the code review process @cedk!

**Comment by vstinner** [comment]
Well done @cedk! I merged your PR. Thanks for your nice enhancement.
